oximeter/db/src/client.rs

```diff
@@ -3338,4 +3680,524 @@ mod tests {
+            .await
+            .expect("Failed to apply one schema upgrade");
+
+        // Check that it actually worked!
+        let body = client
+            .execute_with_body(format!(
+                "\
+            SELECT name, type FROM system.columns \
+            WHERE database = '{test_name}' AND table = 'tbl' \
+            ORDER BY name \
+            FORMAT CSV;\
+        "
+            ))
+            .await
+            .unwrap();
+        let mut lines = body.lines();
+        assert_eq!(lines.next().unwrap(), "\"col0\",\"UInt8\"");
+        assert_eq!(lines.next().unwrap(), "\"col1\",\"UInt16\"");
+        assert_eq!(lines.next().unwrap(), "\"col2\",\"String\"");
+        assert!(lines.next().is_none());
+    }
+
+    #[tokio::test]
+    async fn test_apply_one_schema_upgrade_replicated() {
+        const TEST_NAME: &str = "test_apply_one_schema_upgrade_replicated";
+        let logctx = test_setup_log(TEST_NAME);
+        let log = &logctx.log;
+        let mut cluster = ClickHouseCluster::new()
+            .await
+            .expect("Failed to initialise ClickHouse Cluster");
+        let address = cluster.replica_1.address;
+        test_apply_one_schema_upgrade_impl(log, address, true).await;
+
+        // TODO-cleanup: These should be arrays.
+        cluster
+            .keeper_1
+            .cleanup()
+            .await
+            .expect("Failed to cleanup ClickHouse keeper 1");
+        cluster
+            .keeper_2
+            .cleanup()
+            .await
+            .expect("Failed to cleanup ClickHouse keeper 2");
+        cluster
+            .keeper_3
+            .cleanup()
+            .await
+            .expect("Failed to cleanup ClickHouse keeper 3");
+        cluster
+            .replica_1
+            .cleanup()
+            .await
+            .expect("Failed to cleanup ClickHouse server 1");
+        cluster
+            .replica_2
+            .cleanup()
+            .await
+            .expect("Failed to cleanup ClickHouse server 2");
+        logctx.cleanup_successful();
+    }
+
+    #[tokio::test]
+    async fn test_apply_one_schema_upgrade_single_node() {
+        const TEST_NAME: &str = "test_apply_one_schema_upgrade_single_node";
+        let logctx = test_setup_log(TEST_NAME);
+        let log = &logctx.log;
+        let mut db = ClickHouseInstance::new_single_node(0)
+            .await
+            .expect("Failed to start ClickHouse");
+        let address = SocketAddr::new(Ipv6Addr::LOCALHOST.into(), db.port());
+        test_apply_one_schema_upgrade_impl(log, address, false).await;
+        db.cleanup().await.expect("Failed to cleanup ClickHouse server");
+        logctx.cleanup_successful();
+    }
+
+    #[tokio::test]
+    async fn test_ensure_schema_with_version_gaps_fails() {
+        let logctx =
+            test_setup_log("test_ensure_schema_with_version_gaps_fails");
+        let log = &logctx.log;
+        let mut db = ClickHouseInstance::new_single_node(0)
```

Comment:
Eventually we'll remove single node mode. For tests that don't require a separate test for single-node/replicated, we may want to favour replicated over single node. When the time comes to remove single node functionality, we'll have less code to remove/adapt.

Reply:
IMO, we're still a bit of time from removing single-node entirely, and I prefer having the ability to run tests individually and in parallel in the meantime.

Reply:
I think I might have not explained myself very well :smile: . What I meant was that for tests like this one where it doesn't matter what type of installation it runs against, it may be a good idea to run it against a cluster instead. Basically just swapping `ClickHouseInstance::new_single_node(0)` for `ClickHouseCLuster::new()`.

Anyway, this is probably just a nit. Either way is probably fine